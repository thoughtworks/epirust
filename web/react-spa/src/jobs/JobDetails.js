/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";

import "./job-details.scss"
import { NavItem } from "../common/NavItem";
import { Redirect, Route, Switch } from "react-router-dom";
import GridPage from "../grid";
import PropTypes from 'prop-types';
import { TimeSeriesDeviation, TimeSeries } from "../time-series";

export const JobDetails = ({ simulationId, details }) => {
  const linkPrefix = `/jobs/${simulationId}`;

  const isGridEnabled = details && details.config.enable_citizen_state_messages,
    isFinished = details && details.status === 'finished';

  function renderTabs() {
    return (
      <div className="row">
        <div className="col-8">
          <ul className="nav nav-tabs nav-fill">
            <NavItem name="Time Series" linksTo={`${linkPrefix}/time-series`} activeOnExactMatch={true} />
            {isFinished && <NavItem name="Time Series Deviation" linksTo={`${linkPrefix}/time-series-deviation`} />}
            {isGridEnabled && <NavItem name="Grid" linksTo={`${linkPrefix}/grid`} />}
            <NavItem name="Config" linksTo={`${linkPrefix}/config`} />
          </ul>
        </div>
        <div className="col-4">
          <h4 className="simulation-id">Simulation: {simulationId}</h4>
        </div>
      </div>
    );
  }

  function renderContentForTab() {
    return (
      <Switch>

        <Route exact path={"/jobs/:id/time-series"}>
          <TimeSeries simulationId={simulationId} />
        </Route>

        <Route exact path={"/jobs/:id/time-series-deviation"}>
          <TimeSeriesDeviation simulationId={simulationId} />
        </Route>

        <Route exact path={"/jobs/:id/grid"}>
          {isGridEnabled ? <GridPage /> : <Redirect to={`/jobs/${simulationId}/time-series`} />}
        </Route>

        <Route exact path={"/jobs/:id/config"}>
          {details && <pre>{JSON.stringify(details.config, undefined, 4)}</pre>}
        </Route>
      </Switch>
    );
  }


  return (
    <div className="job-details" style={{ paddingTop: 4 }}>
      {renderTabs()}
      {renderContentForTab()}
    </div>
  )
};

JobDetails.propTypes = {
  simulationId: PropTypes.number.isRequired,
  details: PropTypes.object
};
