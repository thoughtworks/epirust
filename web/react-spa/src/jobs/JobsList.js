/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, { useEffect, useState } from "react";
import './jobs-list.scss'
import { Job } from "./Job";
import { JobDetails } from "./JobDetails";
import { Redirect, useParams } from 'react-router-dom';
import config from "../config";

export const JobsList = () => {
  const { id, view } = useParams();
  const [simulations, updateSimulations] = useState([]);

  useEffect(() => {
    fetch(`${config.API_HOST}/simulation/`)
      .then(res => res.json())
      .then(value => updateSimulations(value.reverse()))
  }, []);

  if (id && !view) {
    return (<Redirect to={`/jobs/${id}/time-series`} />);
  } else {
    const simulationDetails = simulations.find(s => s.simulation_id === parseInt(id));
    return (<div className="row jobs-list">
      <div className="col-3">
        <ul className="list-group scrollable">
          {simulations.map(s =>
            <Job
              key={s.simulation_id}
              simulationId={s.simulation_id}
              status={s.status} />
          )}
        </ul>
      </div>

      <div className="col-9 left-border scrollable">
        {simulationDetails && id && <JobDetails simulationId={parseInt(id)} details={simulationDetails} />}
      </div>

    </div>
    );
  }
}
