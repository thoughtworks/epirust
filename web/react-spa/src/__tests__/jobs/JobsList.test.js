import {render} from "@testing-library/react";
import {JobsList} from "../../jobs/JobsList";
import React from "react";
import {MemoryRouter} from "react-router-dom";
import {act} from "react-dom/test-utils";

describe('JobsList',  function () {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should fetch simulations when id and view is not selected', async function () {
    const mockResponse = Promise.resolve([{simulation_id: 1234, status: "finished"}]);
    const mockJson = jest.fn().mockReturnValueOnce(mockResponse);
    const mockPromise = {json: mockJson};
    jest.spyOn(global, 'fetch')
      .mockImplementation(() => Promise.resolve(mockPromise));
    jest.mock('react-router-dom', () => {return {
      ...jest.requireActual('react-router-dom'),
      useParams: () => ({})
    }});

    let component;
    await act(async () => {
      component = render(<MemoryRouter><JobsList/></MemoryRouter>);
    });

    expect(component.container).toMatchSnapshot();
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(mockJson).toHaveBeenCalledTimes(1);
  });
});