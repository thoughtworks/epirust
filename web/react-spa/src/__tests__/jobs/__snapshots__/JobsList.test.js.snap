// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsList should fetch simulations when id and view is not selected 1`] = `
<div>
  <div
    class="row jobs-list"
  >
    <div
      class="col-3"
    >
      <ul
        class="list-group scrollable"
      >
        <li
          class="list-group-item borderless item-less-padding"
        >
          <a
            class="no-link-formatting"
            href="/jobs/1234"
          >
            <div
              class="card finished"
            >
              <div
                class="card-body card-body-less-padding"
              >
                <div
                  class="job-status"
                >
                  Finished
                </div>
                <div
                  class="minor-details"
                >
                  Simulation Id: 
                  1234
                   
                </div>
              </div>
            </div>
          </a>
        </li>
      </ul>
    </div>
    <div
      class="col-9 left-border scrollable"
    />
  </div>
</div>
`;
